[READY] Syndicate Contracts

Akrilla opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

![ss+(2019-06-09+at+08 42 48)](https://user-images.githubusercontent.com/18756865/59163314-3ad36b00-8af7-11e9-8bc4-4c95f220853e.png)

## About The Pull Request

Introduces Syndicate contracts. Available through the new 20 TC contract kit, traitors can take on contracts for TC payments. The bundle includes a Syndicate styled tablet computer, with a custom "Contract Uplink" program embedded inside the hard drive. Upon launching, the traitor will assign themselves to it, logging in, and accessing the contract page. There's a short bit of flavour text explaining how to properly do the contracts, and from there they can assign themselves to a contract to complete. Only one can be taken on at once, but after completion or after aborting one in progress, they can choose another. Aborting or completing will remove the contract from the list, meaning they'll be unavailable to take on again.

Contracts are generated unique per traitor (currently 6 available to take on), and assigned at the moment they become one. This means that if they decide to take on contracts later on in the round, there's a chance their target may be completely unavailable to them. The system is designed around being something you take on very early on in the round, spending all of your TC for a high risk, at the chance to achieve more than your 20 TC in the late game if you do a good job. Likely, this'll be something picked up by more experienced players for a unique, difficult experience that encourages interesting and exciting gameplay for both the antag and other players. 

Payments are randomised, but weighted around low payouts, with comparatively well paying bonuses should you bring the target alive. Payouts are done through the tablet uplink - directly having raw telecrystals be redeemable after a successful contract. 

Dropoff points are entirely randomised based on areas/rooms on the current station, with the idea being that the player will need to balance all factors into what contract to go for. Who is the target, is it worth the payment, plus difficulty of getting access to the dropoff point etc etc.

The overall theme/goal of this is to make being a traitor feel more badass, and less, "I got the text that lets me murderbone". Feel free to suggest changes or further ideas! 

TODO:
- [x] Actually add the contract kit to the Syndicate uplink. This'll likely include the contract uplink, and some very very cheap items. E.g. chameleon jumpsuit + mask. Maybe randomised each time from a small pool of the very cheap items.

- [x] Some flavour text/feature for when a body is given to the extraction pod.
- [x] Show completed contracts on round end screen. Total TC payout etc etc.
- [x] Some form of interesting feature for targets that were brought back alive
- [x] Small code refactors/text fixes.
- [x] Ship them to Ninja Jail
- [x] Ransom system - allow crew to pay for their crew back. Ransom number generated entirely randomly, done through comms console(?) Alert could be given through the command channel. Kill target when returning if they didn't pay.
- [x] Ship victim back unharmed. If we received a dead body, after a certain amount of time has passed, we bring them back too.

## Extensions

While not in the scope of this PR, I'll likely be adding the same system to an emaggable bounty console with some unique interactions at some point. Much less contracts, less payouts etc, but available with only an emag. Hopefully can give a nice replacement to the null crate that was removed. The code for this is pretty much done, but removed it near the end of finishing the overall feature off since it wasn't well fleshed out. I can focus on adding this after the core is in.

The same is true for a dedicated antag type much like how blood brothers works along side traditional traitor. On higher pop, give a chance for a traitor to start off as a Syndicate contractor. They can be given the uplink and perhaps some more specialised/unique items/weapons to start off with, with their objectives to be to complete x about of contracts etc.

## Changelog
:cl:
add: Syndicate Contracts. Use the new contract uplink to select a contract, and bring the assigned target dead or alive to the designated drop off. Call for the extraction pod and send them off for your TC payment, with much higher rewards for keeping them alive. A high risk, high reward choice for a traitor who wants a real challenge.
add: New 20 TC contract kit - supplies you with your contractor loadout and uplink.
add: Targets successfully extracted will be held for ransom by the Syndicate after their use to them is fulfilled. Central command covers the cost, but they'll be taking a cut out of station funds to offset their loss...
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Akrilla committed:
Emag bounty computer to show contract connection

Akrilla committed:
Corrupted text helper

Akrilla committed:
WIP on corrupted text

Akrilla committed:
Randomly corrupt text when looking at connect screen

Akrilla committed:
Barebones accessible Syndicate bounties screen

Akrilla committed:
Contract barebones and Synd UI work

Akrilla committed:
Further text changes, antag checks

Akrilla committed:
Further UI changes, full contract generation with areas/target

Akrilla committed:
dropoff point checks, and move to uplink item for requesting bounties

Akrilla committed:
WIP contract uplink item

Akrilla committed:
Contract uplink and continued contract improvements

Akrilla committed:
Completed contract uplink, and continued contract work

Akrilla committed:
Droppod WIP

Akrilla committed:
Further droppod work

Akrilla committed:
Mob WIP

Akrilla committed:
Further mob WIP

Akrilla committed:
No mob

Akrilla committed:
Callback for putting in mob

Akrilla committed:
Further WIP with pod checks

Akrilla committed:
Better location picking and signal fix

Akrilla committed:
Further pod changes, WIP of payouts

Akrilla committed:
Overall finished payment/contract system

Akrilla committed:
Code cleanup


Revert changes to bounty console


Remove debug messages

Akrilla committed:
Small refactor

Ledwin1233 commented:
What happens to the alive player afterward?

wesoda25 commented:
This is awesome

Akrilla commented:
> What happens to the alive player afterward?

To the player's point of view, they'll ghost on the tile the pod launches from, but I'm not the biggest fan of this and probably the biggest issue I have overall. I like the idea of somehow bringing the player back into play at some point later in some way if the antag brings them to the pod alive. Maybe just after 2 minutes or something another pod comes down in a random location, bringing them back with some sort of negative consequence. 

If dead however I think just taking the body away is fine.

Tlaltecuhtli commented:
maybe make it so you can get back the corpses if you kill the antag with the contract uplink
like an "rewrite the rocket trajectory to return in  a random place " botton?

Akrilla commented:
> maybe make it so you can get back the corpses if you kill the antag with the contract uplink
> like an "rewrite the rocket trajectory to return in a random place " botton?

I really like that idea yeah, I think that's something that could work.

Shadowflame909 commented:
@Akrilla You could always put the contracted player to the space ninja den.

It needs more use for a place that hostages should get sent to.

knacker48 commented:
While I will miss null crates I do think this is a well fleshed out system and I look forward to trying it out. Though I will say that 20 tc for the option to do this seems like a lot. Would you be willing to bring down the price to maybe around 15 tc so you can still have some tools to carry out your contracts?

Akrilla commented:
> @Akrilla You could always put the contracted player to the space ninja den.
> It needs more use for a place that hostages should get sent to.

Oh shit yeah, 100% I'll do that at the very least.

Shadowflame909 commented:
Hell Yeah. The lore for it is that the syndicate collects hostages/ nanotransen pays for peoples release.

So it pretty much works either way.

Shadowflame909 commented:
I sincerely hope this gets merged, less murder-bone whilst still having to complete those objectives.

All I know is chemistry traitor is going to be a blast.

necromanceranne commented:
Oh fuck it's spy thief but you're hunting the most dangerous game.

Man.

zxaber commented:
The space ninja den is worse than it ghosting you, because it makes you take the effort of ghosting. There is really nothing entertaining about the place. If it was ever changed to give captured players the chance to escape, then it would be a lot of fun, but as is, it's just a waste of time for everyone captured.

Returning living people is neat especially because you'd now have previous targets who know who you are. So disguises and constantly changing your clothes would be the name of the game. A forged ID and later a chameleon kit would be immediate purchases once you get a few contracts done, and you'd be an antag that revolves around removing players for a short period of time to power up rather than just killing.

Shadowflame909 commented:
I mean if you like botany or the kitchen, it has that for you. It just needs more interesting items.

Bet if it had xenobiology it'd be much more interesting.

zxaber commented:
The problem is that in the ninja den you're basically in a vacuum. You might as well be playing on a private server now, nothing you do will affect anyone that matters, and you have no possible way to get back into the game. So you just ghost and take a ghost role or something.

Sending people to the ninja den is effectively murderboning.

necromanceranne commented:
One thing that's been brought up talking about this pr is the difficulty of kidnapping people due to the drag nerf. This makes getting someone from one side of the station to the other without the hand tele something of a chore if not downright impossible without some janky as fuck methods.

My first suggestion would be to actually provide some tools in the kit to make that a bit easier. Fulton packs come to mind if we want to use MGS memes. 

Another would be to take a leaf out of goon's spy thief's book and keep the drop off to a single location shared by all your contracts. Or, alternatively, specific drop offs which range from low risk (somewhere in maint), to high risk (caps office, the brig), and provide a scaling bonus for using that drop off.

HomocidlePancakes commented:
Target should get replaced if it suicides

Shadowflame909 commented:
No, sending people to the Ninja Den is effectively sending them to centcomm.

For all purposes it counts as escaping alive, so you don't have anywhere else to go.

Also if Akrilla goes through with that "Send people back" option then you really wouldn't have to worry about being stuck their either.

It's pretty perfect, and I will say 1000% I prefer doing chef recipes or bar recipes for my fellow captives, then to be erased.

zxaber commented:
If people get sent back afterwards, I totally would not have a problem with it.

The purpose of the game isn't to "escape alive", it's to have fun with the game mechanics and interacting with other players, spiced up by risk of antagonists. So being sent to centcomm thirty minutes before everyone else is a very small consolation. For people that don't much enjoy botany or mindlessly punching eachother in a boxing ring, there's little to really do.

Shadowflame909 commented:
Not liking botany, cooking, or bartending sounds like a you problem.

Syndicate Contracts are supposed to break up the monotony of forced death and murder-bone. Having them kill more people for TC is effectively futile.

As you said, you prefer to instantly be dead without having to press the suicide button. But that button is there.

Just try some of the new things Service has to offer. 

But for me, I prefer to not have another traitor item that ends up requiring more people to die. I'm pretty sure Maintainers are trying to phase out any more additions of "His Grace" (where you require to kill 20 humans with minds to make it evolve) Or E-Swords.

Maybe there's a disconnect in me understanding your argument. But preferring death over just playing out the Service Department till the round-ends sounds like you're already burnt out over what this game has to offer.

zxaber commented:
The service department is meaningless without people to serve. Perhaps your experience is different, but I've never landed in the ninja den and seen anything except ghosted bodies. Sure, mixing drinks is fun, but if there's no one to interact with, then why bother? I've hung around a bit to see people join, complain about how the ninja was a dick, and just ghost.

Me not liking botany isn't about me being burnt out, it's just me not liking botany. I still quite enjoy robotics and engineering (though the former is mostly for the interactions with borgs, other players). I enjoy robotics so much that I often roll assistant if I don't get the robo job and just set up a secondary robotics in maintenance.

My real point is that forcing people into the ninja den is effectively murderboning. You're cut off from the station, isolated to a room that _at best_ allows you to interact with other players in the same situation, assuming any actually stick around. You can't influence the round, you can't enjoy the conflicts, you're dead to the world. If this PR removes people from the den after a short time, then I have no issues with it; you still get to go back to the station eventually and so it's more like a brig sentence.

MMMiracles commented:
let you rescue dead contracts once the traitor is detained from the contract menu, beam the living ones back after a short time with the fluff it being paid ransom or being brainwashed to be pro-syndicate now.

spider den is gay and is arguably worse than just ghosting the player and deleting their body though.

zxaber commented:
Brainwashing would lead to them being antags (or at least encouraged to assist the specific traitor that captured them), which ultimately means you're getting TC and also building an army of bodyguards. It sounds funny, but also like superpowered cult.

necromanceranne commented:
Why not brainwashed to forget the event even happened? Why does it need to be pro-syndicate?

wesoda25 commented:
rules dont give a fuck if you suicide roundstart, you can abort a mission anyways

plapatin commented:
cant wait to see romerol bought every other round

GuyonBroadway commented:
Add holopad to ninja den so people can call back.

Add a way for the station to pay bounties to bring alive people home.

Akrilla commented:
> While I will miss null crates I do think this is a well fleshed out system and I look forward to trying it out. Though I will say that 20 tc for the option to do this seems like a lot. Would you be willing to bring down the price to maybe around 15 tc so you can still have some tools to carry out your contracts?
My worry is balancing something that can give you TC, and I’d rather this be in the game as something too weak than OP, I can always change things after it’s been in for a while. Saying that though, the kit will likely come along with some randomised low cost traitor gear to help justify the price.

Akrilla commented:
> cant wait to see romerol bought every other round

I don’t think you realise how difficult these contracts are and how little they really pay out comparatively...

Tlaltecuhtli commented:
why not make it cost 15TC so you are forced to use the weak tier syndi tools (dagger,foam pistol,pen)

FloranOtten commented:
If you're going to go with the 'return people after a short time' option, they need to forget. Imagine going through all the effort to kidnap people only for them to return a short while later screaming your name.

Brainwashing to be syndicate is a terrible idea. Then why even resist to kidnapping? Free antag! Add mindshields to the mix and you've got a small scale conversion gamemode.

Michael-Ashfield commented:
I do quite like the idea of ransoms, gives the economy some actual use. Perhaps victims could be given a 5-10min timeout before the station gets an announcement asking to pay the ransom (if the player is still alive), which they can pay via a comms console, if they pay the victim is bluespaced back with their mind wiped.
I can totally see departments banding together to save their heads and members, such as cargo raiding every welding fuel tank and crate in the station to save the QM.

This gives some flavour behind the contracts, as the syndicate profit from them and can justify giving the agent more telecrystals for expensive syndie gear.

The cost of the ransom would have to be balanced so it doesn't make it too easy to get someone back, or too expensive to bring back less 'valuable' crew such as assistants and curators. Somewhere around the 4000 mark, maybe scaling on rank (captain would be 8k).

This gives a chance of return, a reason not to immediately ghost, crew teamwork, use of economy and gives some safety for the agent doing the contact (so they don't get immediately exposed).

SgtxRootbeer commented:
Working with ransom, sounds rather interesting.
I'd personally prefer a more hugboxy approach considering how hard medical is being nerfed currently,

maybe something like to get your TC, you're required to bring them to the pod alive. (makes the traitor wanna use both lethal for defense, against crew that the station will eventually rescue, and non-lethal against their target, or do something sneaky involving secret cloning.) Hostage bounties should be auto-paid, authorized by nanotrasen brass, coming from the person's personal account first, then their dept account, and finally station's budget. maybe give the comms console a way to reject hostage deals if the subject was someone the station REALLY didn't care for. to compensate for the issue of returning people, the kit should come with a randomly selected stealth package. (1: voice-changer and agent ID. 2: chameleon kit and mulligan. 3: no-slips and holo-projector.)

I don't much care for the murdering people, because its sorta useless to deliver a dead body to your superiors if they originally planned on ransoming your target. This also fixes the issue where a person who would have otherwise just killed their 1 target, is now going to kill 7 people instead, because the contract is motivating them.

Akrilla commented:
From the comments here, I think what I'll likely decide to do is have the kit still 20 TC, but with a guaranteed chameleon jumpsuit, mask and mulligan. Outside those, there's two random rolls of very cheap items from a set of items I'll make. I suppose that "small stealth bundle" would work too.

I think I like the ransom idea the most, it'd be interesting to see the crew rally around to try and get the funds to get their department head back. They'll have "forgot" their experience, so it becomes an admin issue if they rat out the traitor that kidnapped them. I'll maybe start theming things over to that idea.

zxaber commented:
Will they need to "forget" the experience if we're including disguise equipment too? Honestly, even without the extra equipment, the traitor could simply grab a random jumpsuit from the dorms and a gasmask and just do their first or second contract as unknown.

The problem with random is that I am willing to bet no one bothers with it. You see this same issue with the piles of dead bodies that lie outside cloning. No one will make an effort to get the person back (except with metafriends, I suppose), unless there's a very good reason, like the person being the clown. So we're in the same boat in that you're almost certainly removed from the round, but now you can't even observe it or take a ghost role because of the slight chance you _might_ get rescued.

Auto-emptying accounts would be a decent idea. Take from the captured first if we must, it certainly would be an extra incentive not to be caught, and then take from the department, as a way to make departments as a whole invested in catching the person or preventing further abductions. You could even have contracts tend to pick several people from one department, to make that department band together in the name of safety.

Michael-Ashfield commented:
I don't see a reason why corpses or even freshly cloned victims can't be returned after their ransom is paid by the station. This lets people ghost or be sent up dead and still have a chance to get back into the round.
People being dead in a round isn't a massive problem that everything should be designed around avoiding, in any other situation the target would have been killed by a traitor and just spaced, cremated or any other non revivable situation. If a barely armed traitor can capture you in some way, drag you to a probably restricted area, keep you there while they send you up using a very valid tablet and get away with it all, then that's fair, especially if you still have some chance to be recovered.
Having a big syndie announcement telling everyone there is a ransom is sure enough to get some peoples attention and could lead to some fun RP.

Akrilla commented:
> Auto-emptying accounts would be a decent idea. Take from the captured first if we must, it certainly would be an extra incentive not to be caught, and then take from the department, as a way to make departments as a whole invested in catching the person or preventing further abductions. You could even have contracts tend to pick several people from one department, to make that department band together in the name of safety.

Taking from the idea of having contracts weigh on departments - having contract bundles? Capture/kill and bring 3 heads of staff, bring 3 cargo members, etc etc. Then get one big lump sum after the entire package is complete. Maybe for a future PR.

datorangebottle commented:
> The space ninja den is worse than it ghosting you, because it makes you take the effort of ghosting.

What in the world makes you think that ghosting/suiciding takes effort? It's what, four buttons? It takes all of ten seconds. Five if you're going fast.

Anyway, I like this PR in its entirety. It rewards skilled traitors with extra TC for putting in effort. I dislike any and all attempts by you lot to try and change it to make it come back and bite the traitor in the ass for doing an excellent job. "Durr you can get the person back, so you literally are required to spend some of your hard-earned TC on a chameleon kit!!!1!" How is this an interesting conflict or interaction? How does this encourage people to actually use this system? Hint: it's not and it doesn't.

Including 'free' traitor items kinda defeats the purpose, doesn't it? It makes the contract even more valuable than it should be.

zxaber commented:
You entirely missed the point of that sentence.

Akrilla commented:
>Including 'free' traitor items kinda defeats the purpose, doesn't it? It makes the contract even more valuable than it should be.

While this is definitely meant to be some sort of "hardcore" mode of playing traitor, with the hook being that if you're up for it you can get rewarded nicely in the late game, I want to at least give some sort of extra value. Most players I'd say wouldn't be able to get their TC back, and even the better antag players are going to struggle. Probably going to give them the very basics of the chameleon kit, a custom syndicate space suit (just going to be colored differently from the 4 TC one) then maybe two entirely random low costing traitor items. 

20 TC bundles tend to be overall be worth more than the 20 TC you pay, with the trade off being not knowing what you're going to get. This I feel works the same with giving set low cost items, with an item you basically bought at very high risk.

anconfuzedrock commented:
if the contracts are just kidnapping this is prolly gonna get very bland very fast, at minimum it could have the already existing stealy traitor objs

Akrilla commented:
>if the contracts are just kidnapping this is prolly gonna get very bland very fast, at minimum it could have the already existing stealy traitor objs

I'll be actively expanding this once the core is in, but I don't really agree it'll really be that bland with kidnapping.

> if we do have ransoms and them being bought though then I very much so think them "forgetting" the experience is a bad idea.

I want to encourage traitors to bring them back alive, if it's too difficult they're just going kill their target regardless of the bonus.

fludd12 commented:
This may be a somewhat contentious idea, but I'd actually suggest that this system be available by default for traitors, and to significantly reduce or even neuter their start-of-round telecrystals. This way, traitors can accomplish as many or as few side-objectives as they choose to gain better resources to complete their main objectives, kind of like Dishonored's system of allowing multiple paths to make the primary assassination target easier to get to.

This, of course, can be a decision made further down the line, after this PR is done, but I figured I'd pose the idea to you!

Akrilla commented:
I'm a bit biased since I don't like murderbone, but I like that idea a lot. Only problem is that I think a lot of other people wouldn't be too keen since some play Synd to buy what they like and mess about. At the very least, a version of that would be part of the contractor antag type - someone who starts with no TC, but with their kit and unique starter equipment.

fludd12 commented:
You could have more contracts than just kidnappings, and possibly make contracts refresh upon completion, and after ~5-10 minutes upon abandoning them. That way, the people who want to buy what they like could take several, low-risk-low-reward contracts to get the meme kit they need to faff about and ignore greentext, while more determined traitors would either go without TC, or do the contracts they need to complete their objectives, and take contracts after as bonus rewards.

Perhaps you could have a round-end statistic for black market contracts completed, and have some fancy highlighting of the traitor's name for certain amounts of contracts completed.

I do like the idea of a contractor antagonist, though I do think that the normal scheme of traitor is due for a bit of shaking-up lately, since it's gotten somewhat stagnant.

Akrilla committed:
Merge remote-tracking branch 'tgstation/master' into syndicate-contracts-nobountyconsole

Tlaltecuhtli commented:
can you add the ability for other traitors to create new contracts? (they pay with their TCs and can choose the target)

Akrilla commented:
> can you add the ability for other traitors to create new contracts? (they pay with their TCs and can choose the target)

As in, non contractors paying for current contractors to kill someone? That'd be really cool, but I'd worry about it being metagamed, or in what way to best do it. If you select a target, how do you know there's even someone that has the kit? If you tell them there's nobody, or by letting them choose a target, you tell them something they maybe shouldn't be able to know.

Akrilla committed:
Add kit with set and randomised items.

Akrilla commented:
Finally actually decided what items to add and made it jumpsuit, mask, space suit and two random items from a pool of cheap items. I feel like that’s a pretty nice mix, and still balances things nicely. Traitor starts off weak, but some interesting items to play with.

Akrilla committed:
Fix box

qbmax32 commented:
unironically make this default rather then an item

Akrilla committed:
Very minor flavor text for inserting into pod, plus roundend text

anconfuzedrock commented:
I’m still certain that them “forgetting” is the worst idea ever and at best you’ll get the kidnapped and released warden casually watching their kidnapper on security cameras. If you don’t want anyone to die, just make the reward for a murder 1 tc or something unprofitable. 

And on the “make this default” That’d be such a major change to our core antagonist that we should at least try this as an item first.

zxaber commented:
Yeah, I much like the idea of the abduct-traitor using disguise gear or at least dorms jumpsuits and masks to hide his identity from the people he abducts. It's a gameplay/strategy solution rather than a admin one.

Akrilla commented:
I'm sold tbh, the contract kit already comes with the chameleon jumpsuit and mask anyway, I'll do that. Though, I'm likely to reduce the "bringing them dead" rewards, and slightly increasing the reward for bringing them alive. Or, reduce the bringing them dead, keep alive reward the same, but throw in a free agent card.

Akrilla committed:
Pod return sound

Akrilla committed:
Contract uplink slight styling changes

Akrilla committed:
We don't keep giving telecrystals until they literally can't hold anymore

Akrilla committed:
Ship them off to ninja base instead

Akrilla committed:
Nerf bring them dead, but give a nice boost for bring them alive

Overall this is a nerf, so we also throw in an agent card to the bundle

Akrilla committed:
Ship them back

anconfuzedrock commented:
Yeah, reduced rewards for corpses is probably all that's needed. After all, if it's unprofitable to buy this and murder people, people will just murder without buying this.

Akrilla committed:
Minor text changes and slight TC change

Akrilla committed:
Always give the black and red variant space suit

Akrilla committed:
Error sounds

Tlaltecuhtli commented:
what happens to the abductee worn items?

Akrilla commented:
Kept

Akrilla committed:
Dropoff area locator

There's sometimes duplicates, i.e. security office - but only one of them will work. This shows you where to go.

Akrilla committed:
Small fixes

Akrilla committed:
Bugfix, further changing of TC

Akrilla committed:
Ransom wip

Akrilla committed:
Ransom system

Akrilla committed:
Few changes

Akrilla committed:
Properly pay for ransoms and appropriate logic for dropoffs

Akrilla committed:
Terminal sounds

Akrilla committed:
Talk on command channel when ransom is given, or if they were too late

Akrilla committed:
Change to announce system

zxaber commented:
I still bet that making the crew manually pay for the ransoms is effectively killing 90% of people that get taken. Unless it's the Captain, clown, or someone that's been personally helping the captain all round, no one will probably bother.

If you want to get under people's skin with the capture thing, maybe have NT auto-pay the bounty to get the crew member back, and then penalize the Captain and all security members with a fine for "not preventing the abduct in the first place".

Akrilla commented:
Yeah fair enough, I think it would be even more challenging for the traitor if you get the entire crew pissed at you, ending up having cargonia etc after you

Akrilla committed:
Announcements/auto pay ransom

Akrilla committed:
Proper wait

Akrilla commented:
Feature complete, ready to merge after review

Akrilla committed:
Name pods properly

Tlaltecuhtli commented:
not sure of the pod works but is it possible to abuse it to send stuff like other people on the tile away? (or tracking beacons etc)

Akrilla commented:
> not sure of the pod works but is it possible to abuse it to send stuff like other people on the tile away? (or tracking beacons etc)

ugh

Akrilla committed:
Don't send blacklisted items

Akrilla committed:
fix

AnturK commented on code/game/objects/items/storage/uplink_kits.dm:
```diff
@@ -180,6 +180,54 @@
+	var/first_num = rand(1, item_list.len)
+	var/second_num = rand(1, item_list.len)
+
+	while (second_num == first_num)
```
You can just pick_n_take from the list,since the list is not static/shared.

AnturK commented on code/modules/antagonists/traitor/datum_traitor.dm:
```diff
@@ -356,6 +369,19 @@
+	if (completed_contracts > 1) 
+		pluralCheck = "contracts"
+	
+	result += "Completed <span class='greentext'>[completed_contracts]</span> [pluralCheck] for a total of \
```
Probably should only display this if they did any at all.

AnturK commented on code/game/gamemodes/objective.dm:
```diff
@@ -1054,3 +1054,24 @@ GLOBAL_LIST_EMPTY(possible_items_special)
 		var/datum/objective/X = T
 		GLOB.admin_objective_list[initial(X.name)] = T
+
+/datum/objective/contract
```
I'm confused why you need this objective type if pretty much everything is handled by contract datum and you do not assign this objective at all.

Akrilla replied:
While not being a "traditional" objective, it still matches a lot of the criteria. We have a target, an objective owner (contract owner), and it makes use of the target find function. It also will be further extended when this'll likely be used for the contractor antag type in the future.

AnturK commented on code/modules/antagonists/traitor/datum_traitor.dm:
```diff
@@ -23,10 +27,19 @@
 	RegisterSignal(owner.current, COMSIG_MOVABLE_HEAR, .proc/handle_hearing)
 	..()
 
+/datum/antagonist/traitor/proc/create_contracts()
```
You could probably delay generating these until  they first access the contract interface.

AnturK commented on code/modules/modular_computers/file_system/programs/antagonist/contract_uplink.dm:
```diff
@@ -0,0 +1,133 @@
+			var/turf/dropoff_turf 
+			data["current_location"] = "[get_area_name(curr, TRUE)]"
+			
+			for (var/content in traitor_data.current_contract.contract.dropoff.contents)
```
Areas have z property that's lowest z among their turfs so you don't really need to iterate here from what i see.

Akrilla replied:
This loop is for finding the first instance of a turf inside the area, since .contents gives turfs as well as obj. Is there a better way to get that from what you're saying?

AnturK replied:
Just `area_instance.z` is enough.

AnturK reviewed (changes requested):
Some improvements.

Akrilla committed:
Improvements

Akrilla committed:
Announcements are less huge...

Akrilla committed:
Not a file...

Akrilla committed:
Improvement

Akrilla committed:
Drop them not in a wall

timothymtorres commented:
I have a major suggestion.

This should send the hostages to the syndie outpost/base in a jail cell.  I know that’s be a lot of work but it’d definitely be much more interesting for all players involved.

123chess456 commented:
I would suggest adding a voice changer and agent ID if they're not already in the kit. That way your target won't be able to identify you once they get ransomed back.

habader commented:
do you think this could be its own game mode where all tators are forced to do contracts

Akrilla commented:
> I have a major suggestion.
> 
> This should send the hostages to the syndie outpost/base in a jail cell. I know that’s be a lot of work but it’d definitely be much more interesting for all players involved.

Does Syndie base already have cells, or are you saying they need to be mapped in? I agree though, would be cool to have a holding cell down there. Maybe as an extension to this.

> I would suggest adding a voice changer and agent ID if they're not already in the kit. That way your target won't be able to identify you once they get ransomed back.

Contract uplink, space suit, cham jumpsuit and mask, agent ID, and two random very cheap items is the bundle currently.

> do you think this could be its own game mode where all tators are forced to do contracts

See extensions heading in the PR.

Akrilla commented:
Not entirely ready to be merged yet, need to fix a couple things. Stopping people getting stamina weapons out of ninja jail, reducing a bit of the effects that you get when captured etc

Akrilla committed:
Halves eye blur and styles payment vend

Akrilla committed:
No memes allowed

Akrilla commented:
Effects of getting kidnapped have been more than halved (it was kinda dumb), and nothing can come out or in of the jail. We give items back on return.

timothymtorres commented:
I don’t believe either syndie base has jail cells.  They would need to be mapped in.  Also there would be several things that need to be done.

Traitors would need a way to contact the syndie base.  They would also need a way to teleport their targets (launchpad? Fulton beacon?) to the base.  Right now the syndie base has a way to access the PDA server and cameras but that’s it.

Not every round has a syndie base spawn.  Also since the syndies at the base are mob spawner roles, it’s also possible that nobody would be active at the base.

Maybe make some type of computer where they need to activate stuff before any traitors can interact with the base?  That’s be my suggestion.
